### DISCUSSION TITLE
Seeking clarification on the distinction between Points, Nodes, Vertices, and Quadrature Points in libmesh

### DISCUSSION THREAD

**Discussion by hityyds** [idaholab/moose#26838] [Q&A General] [answered]
I am working on calculating the center point of an element's side by averaging the coordinates of each vertex. I have been exploring the information available regarding libmesh::Elem for this purpose. However, I am currently confused about the differences among Point, Node, Vertex, and Quadrature Point.

My current understanding is as follows:
    A Point solely contains coordinate information.
    A Node builds upon a Point and is associated with degrees of freedom; every Node contains a Point.
    In MOOSE, each Quadrature Point corresponds to unknown variables (i.e. degrees of freedom), suggesting it should be linked to a Node.
    The number of Vertices is determined by the type of the element, representing physically existing points.

The confusion arises from the fact that both Nodes and Vertices seem to use the same indexing scheme when accessing their positions via the `elem->point(n)` function. 
```
 Point Elem::vertex_average() const
 {
   Point cp;
 
   const auto n_vertices = this->n_vertices();
 
   for (unsigned int n=0; n<n_vertices; n++)
     cp.add (this->point(n));
 
   return (cp /= static_cast<Real>(n_vertices));
 }


std::unique_ptr<MeshBase>
CoarseMeshExtraElementIDGenerator::generate()
{
    ......

    // Check if the fine elem is nested in the coarse element
    for (unsigned int n = 0; n < elem->n_nodes(); n++)
    {
      // Get the node: we need to manually move it towards the centroid to
      // ensure that nothing weird happes due to round-off
      Point current_node = elem->point(n);

      ......
}
```
This gives me the impression that Vertices and Nodes share the same index set, which seems contradictory given that their counts are different.

Furthermore, my understanding suggests that each element has its unique set of Quadrature Points while Nodes can be shared across elements. This seemingly contradicts my earlier notion that each Quadrature Point corresponds to a single Node.

I would greatly appreciate your help in clarifying these concepts and how they relate to finding the center point of an element's side. Thank you in advance for your attention and assistance.

**Comment by lindsayad** [comment]
My advice would be to use [Elem::build_side_ptr](https://mooseframework.inl.gov/docs/doxygen/libmesh/classlibMesh_1_1Elem.html#a1bf88879d9f98fe51372e11a77ade828) and then just call `vertex_average` or `true_centroid` with the resulting element

**Reply by hityyds** [reply]
Thank you for your advice. I had not come across this function before, and it appears to be exactly what I am looking for.

**Comment by lindsayad** [comment, marked as answer]
But returning to your question:

> A Point solely contains coordinate information.

Yes

> A Node builds upon a Point and is associated with degrees of freedom; every Node contains a Point.

Yes, although I think it would be slightly more accurate to say a `Node` is a `Point`

> In MOOSE, each Quadrature Point corresponds to unknown variables (i.e. degrees of freedom), suggesting it should be linked to a Node.

Quadrature points generally have no link to nodes. They are associated with elements. [This Wikipedia article](https://en.wikipedia.org/wiki/Gaussian_quadrature) describes our most commonly used quadrature rule. Quadrature points are used for numerical integration. They don't have any link to variable's degrees of freedom

> The number of Vertices is determined by the type of the element, representing physically existing points.

This is true. I would probably not capitalize vertices because there is no vertex class type in libMesh/MOOSE whereas Point and Node are class types. Vertices are a subset of an element's Nodes. In the case of a Quad4 element, all the Nodes are vertices. However, with a Quad9 element there are also Nodes at the mid-points of element edges and also a Node at the center of the element. 4 vertex nodes + 4 mid-edge nodes + center node = 9 nodes in total.

**Reply by hityyds** [reply]
Thank you for your detailed response. 

> Vertices are a subset of an element's Nodes. In the case of a Quad4 element, all the Nodes are vertices. However, with a Quad9 element there are also Nodes at the mid-points of element edges and also a Node at the center of the element. 4 vertex nodes + 4 mid-edge nodes + center node = 9 nodes in total.

I understand that vertices is part of nodes, but what intrigues me is why the `vertex_average()` function uses `this->point(n)` to access vertex data, which gives the impression that points and vertices are either entirely synonymous or, at least, the first `n_vertices` points correspond directly to vertices.
```
 Point Elem::vertex_average() const
 {
   Point cp;
 
   const auto n_vertices = this->n_vertices();
 
   for (unsigned int n=0; n<n_vertices; n++)
     cp.add (this->point(n));
 
   return (cp /= static_cast<Real>(n_vertices));
 }
```

**Reply by lindsayad** [reply]
> at least, the first n_vertices points correspond directly to vertices.

this is exactly correct. libMesh stores an element's nodes such that the first `n_vertices()` of nodes are indeed the vertices, with mid-edge nodes, etc. following the vertices

**Reply by hityyds** [reply]
I get it, thank you again for your help.

**Reply by lindsayad** [reply]
thanks for always having such good questions!
